### PR TITLE
Wait for container cleanup before deletion

### PR DESCRIPTION
**Pull request by dashpole**
We should wait to delete pod API objects until the pod's containers have been cleaned up. See issue: #50268 for background.

This changes the kubelet container gc, which deletes containers belonging to pods considered "deleted".
It adds two conditions under which a pod is considered "deleted", allowing containers to be deleted:
Pods where deletionTimestamp is set, and containers are not running
Pods that are evicted

This PR also changes the function PodResourcesAreReclaimed by making it return false if containers still exist.
The eviction manager will wait for containers of previous evicted pod to be deleted before evicting another pod.
The status manager will wait for containers to be deleted before removing the pod API object.

/assign @vishh

### CODE REVIEW

**Comment by dashpole** [comment]
cc @kubernetes/sig-node-pr-reviews

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -741,6 +742,22 @@ func (kl *Kubelet) podIsTerminated(pod *v1.Pod) bool {
 	return status.Phase == v1.PodFailed || status.Phase == v1.PodSucceeded || (pod.DeletionTimestamp != nil && notRunning(status.ContainerStatuses))
 }
 
+// podIsDeleted returns true if the pod is deleted.  For the pod to be deleted, either:
```
Comment (resouer) [inline pkg/kubelet/kubelet_pods.go]: nits: IsPodDeleted ...

**Comment by dashpole** [comment]
lgtm removed because of rebase.  This should be in the 1.8 milestone

**Comment by calebamiles** [comment]
- added to the 1.8 milestone.
- manually applied `lgtm` after removal for rebase

**cc: @kubernetes/kubernetes-release-managers**

**Comment by dashpole** [comment]
rebased

**Comment by dims** [comment]
re-applying lgtm after rebase, originally lgtm was from @dchen1107

**Pull request merged**